File: clients/java/src/main/java/io/volant/Scram.java
package io.volant;

import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;
import javax.crypto.Mac;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Client-side SCRAM proof computation (v0.46 SHA-256; v0.238 SHA-512).
 *
 * <p>Matches {@code crates/volant-client/src/scram.rs} (no channel binding).
 */
final class Scram {
    private Scram() {}

    static String generateClientNonce() {
        byte[] buf = new byte[16];
        new SecureRandom().nextBytes(buf);
        return Base64.getEncoder().encodeToString(buf).replace(',', 'A');
    }

    static final class Proof {
        final byte[] clientProof;
        final byte[] serverSignature;

        Proof(byte[] clientProof, byte[] serverSignature) {
            this.clientProof = clientProof;
            this.serverSignature = serverSignature;
        }
    }

    static Proof clientProofAndServerSig(
            String username,
            String password,
            String clientNonce,
            String combinedNonce,
            byte[] salt,
            int iterations) {
        return proofFor(
                username,
                password,
                clientNonce,
                combinedNonce,
                salt,
                iterations,
                256,
                "PBKDF2WithHmacSHA256",
                "HmacSHA256",
                "SHA-256");
    }

    static Proof clientProofAndServerSigSha512(
            String username,
            String password,
            String clientNonce,
            String combinedNonce,
            byte[] salt,
            int iterations) {
        return proofFor(
                username,
                password,
                clientNonce,
                combinedNonce,
                salt,
                iterations,
                512,
                "PBKDF2WithHmacSHA512",
                "HmacSHA512",
                "SHA-512");
    }

    private static Proof proofFor(
            String username,
            String password,
            String clientNonce,
            String combinedNonce,
            byte[] salt,
            int iterations,
            int bits,
            String pbkdfAlg,
            String hmacAlg,
            String digestAlg) {
        try {
            byte[] salted = pbkdf2(password, salt, iterations, bits, pbkdfAlg);
            byte[] clientKey = hmac(hmacAlg, salted, "Client Key".getBytes(StandardCharsets.UTF_8));
            byte[] storedKey = digest(digestAlg, clientKey);
            byte[] serverKey = hmac(hmacAlg, salted, "Server Key".getBytes(StandardCharsets.UTF_8));
            String authMessage = buildAuthMessage(username, clientNonce, combinedNonce, salt, iterations);
            byte[] authBytes = authMessage.getBytes(StandardCharsets.UTF_8);
            byte[] clientSignature = hmac(hmacAlg, storedKey, authBytes);
            byte[] proof = new byte[bits / 8];
            for (int i = 0; i < proof.length; i++) {
                proof[i] = (byte) (clientKey[i] ^ clientSignature[i]);
            }
            byte[] serverSig = hmac(hmacAlg, serverKey, authBytes);
            return new Proof(proof, serverSig);
        } catch (GeneralSecurityException e) {
            throw new ProtocolException("scram crypto failed: " + e.getMessage(), e);
        }
    }

    static boolean signaturesEqual(byte[] a, byte[] b) {
        return MessageDigest.isEqual(a == null ? new byte[0] : a, b == null ? new byte[0] : b);
    }

    private static byte[] pbkdf2(String password, byte[] salt, int iterations, int bits, String alg)
            throws GeneralSecurityException {
        PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, iterations, bits);
        SecretKeyFactory factory = SecretKeyFactory.getInstance(alg);
        try {
            return factory.generateSecret(spec).getEncoded();
        } finally {
            spec.clearPassword();
        }
    }

    private static byte[] hmac(String alg, byte[] key, byte[] data) throws GeneralSecurityException {
        Mac mac = Mac.getInstance(alg);
        mac.init(new SecretKeySpec(key, alg));
        return mac.doFinal(data);
    }

    private static byte[] digest(String alg, byte[] data) throws GeneralSecurityException {
        return MessageDigest.getInstance(alg).digest(data);
    }

    private static String buildAuthMessage(
            String username, String clientNonce, String combinedNonce, byte[] salt, int iterations) {
        String clientFirstBare = "n=" + username + ",r=" + clientNonce;
        String serverFirst =
                "r=" + combinedNonce + ",s=" + Base64.getEncoder().encodeToString(salt) + ",i=" + iterations;
        String clientFinalWoProof = "c=biws,r=" + combinedNonce;
        return clientFirstBare + "," + serverFirst + "," + clientFinalWoProof;
    }
}

File: clients/go/scram.go
package volant

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"hash"
	"strings"
)

// generateClientNonce returns 16 random bytes, standard Base64, ',' → 'A'.
func generateClientNonce() (string, error) {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return strings.ReplaceAll(base64.StdEncoding.EncodeToString(buf), ",", "A"), nil
}

// ClientProofAndServerSig matches crates/volant-client/src/scram.rs (SHA-256).
func ClientProofAndServerSig(username, password, clientNonce, combinedNonce string, salt []byte, iterations uint32) (proof, serverSig []byte, err error) {
	return clientProofAndServerSig(sha256.New, 32, username, password, clientNonce, combinedNonce, salt, iterations)
}

// ClientProofAndServerSigSHA512 is the v0.238 SHA-512 proof (64-byte digest).
func ClientProofAndServerSigSHA512(username, password, clientNonce, combinedNonce string, salt []byte, iterations uint32) (proof, serverSig []byte, err error) {
	return clientProofAndServerSig(sha512.New, 64, username, password, clientNonce, combinedNonce, salt, iterations)
}

func clientProofAndServerSig(h func() hash.Hash, dklen int, username, password, clientNonce, combinedNonce string, salt []byte, iterations uint32) (proof, serverSig []byte, err error) {
	if iterations == 0 {
		return nil, nil, fmt.Errorf("scram iterations must be > 0")
	}
	salted := pbkdf2HMAC(h, []byte(password), salt, int(iterations), dklen)
	clientKey := hmacHash(h, salted, []byte("Client Key"))
	storedKey := digestHash(h, clientKey)
	serverKey := hmacHash(h, salted, []byte("Server Key"))
	authMessage := buildAuthMessage(username, clientNonce, combinedNonce, salt, iterations)
	clientSig := hmacHash(h, storedKey, []byte(authMessage))
	proof = make([]byte, dklen)
	for i := 0; i < dklen; i++ {
		proof[i] = clientKey[i] ^ clientSig[i]
	}
	serverSig = hmacHash(h, serverKey, []byte(authMessage))
	return proof, serverSig, nil
}

func pbkdf2HMAC(h func() hash.Hash, password, salt []byte, iterations, keyLen int) []byte {
	hLen := h().Size()
	nBlocks := (keyLen + hLen - 1) / hLen
	out := make([]byte, 0, nBlocks*hLen)
	var blockNum [4]byte
	for i := 1; i <= nBlocks; i++ {
		binary.BigEndian.PutUint32(blockNum[:], uint32(i))
		u := hmacHash(h, password, append(append([]byte{}, salt...), blockNum[:]...))
		t := append([]byte(nil), u...)
		for j := 1; j < iterations; j++ {
			u = hmacHash(h, password, u)
			for k := range t {
				t[k] ^= u[k]
			}
		}
		out = append(out, t...)
	}
	return out[:keyLen]
}

func hmacHash(h func() hash.Hash, key, data []byte) []byte {
	m := hmac.New(h, key)
	_, _ = m.Write(data)
	return m.Sum(nil)
}

func digestHash(h func() hash.Hash, data []byte) []byte {
	d := h()
	_, _ = d.Write(data)
	return d.Sum(nil)
}

func buildAuthMessage(username, clientNonce, combinedNonce string, salt []byte, iterations uint32) string {
	clientFirstBare := "n=" + username + ",r=" + clientNonce
	serverFirst := fmt.Sprintf("r=%s,s=%s,i=%d", combinedNonce, base64.StdEncoding.EncodeToString(salt), iterations)
	clientFinalWoProof := "c=biws,r=" + combinedNonce
	return clientFirstBare + "," + serverFirst + "," + clientFinalWoProof
}
